### PR TITLE
feat(outputs.timestream): Support ingesting multi-measures

### PR DESCRIPTION
**Pull request by lohithshetty**
Fixes: #10493

### Required for all PRs:

- [x] Updated associated README.md.
- [x] Wrote appropriate unit tests.
- [x] Pull request title or commits are in [conventional commit format](https://www.conventionalcommits.org/en/v1.0.0/#summary)

Feature: support ingesting multi-measures to timestream

The feature controlled using the use_multi_measure_records field which is set to false by default for backward compatibility.

### CODE REVIEW

**Comment by reimda** [comment]
Hi Lohith, thanks for the PR. Could you update the sample config to include the new settings? https://github.com/influxdata/telegraf/blob/master/docs/developers/SAMPLE_CONFIG.md#style

Also update readme.md to describe how the settings are useful and maybe link to timestream documentation?

**Comment by srebhan** [comment]
Hey @lohithshetty, how does this code relate to the other two PRs (#11094 and #10952) with the exactly same PR title? Can we close those?

**Comment by lohithshetty** [comment]
Hi reimda@, apologies for the delayed response. 
I've updated readme.md to include details about multi-measure feature.

**Comment by lohithshetty** [comment]
Hi @[srebhan](https://github.com/srebhan) the other two PRs (https://github.com/influxdata/telegraf/pull/11094 and https://github.com/influxdata/telegraf/pull/10952) can be closed as this PR covers the same change.

File: plugins/outputs/timestream/timestream.go
Comment (srebhan) [inline plugins/outputs/timestream/timestream.go:18]: Can we please keep this newline to separate 3rd party imports from telegraf internal ones?

File: plugins/outputs/timestream/timestream.go
Comment (srebhan) [inline plugins/outputs/timestream/timestream.go:110]:
```suggestion
			return fmt.Errorf("in %q mapping mode, with multi-measure enabled, key MeasureNameForMultiMeasureRecords is invalid", MappingModeMultiTable)
```

File: plugins/outputs/timestream/timestream.go
Comment (srebhan) [inline plugins/outputs/timestream/timestream.go:125]: Why is this different from the code in lines 104 to 110?

File: plugins/outputs/timestream/timestream.go
Comment (srebhan) [inline plugins/outputs/timestream/timestream.go:247]: Errors should start with lowercase letters according to convention. Furthermore, sending the metric is retried in the next flush cycle, so ` Skipping metric!` is misleading I think.

File: plugins/outputs/timestream/timestream.go
Comment (srebhan) [inline plugins/outputs/timestream/timestream.go:460]: How about
```suggestion
	multiMeasureName := t.MeasureNameForMultiMeasureRecords
	if t.MappingMode == MappingModeSingleTable {
		multiMeasureName = point.Name()
	}
```

File: plugins/outputs/timestream/timestream.go
Comment (srebhan) [inline plugins/outputs/timestream/timestream.go:471]: Is this an error or rather a warning?

**Review by srebhan — COMMENTED** [review]
Thanks for updateing the PR @lohithshetty! I have some minor comments in the code...

File: plugins/outputs/timestream/timestream.go
Comment (lohithshetty) [reply plugins/outputs/timestream/timestream.go:247]: Updated, thank you

File: plugins/outputs/timestream/timestream.go
Comment (lohithshetty) [reply plugins/outputs/timestream/timestream.go:18]: Makes sense

File: plugins/outputs/timestream/timestream.go
Comment (lohithshetty) [reply plugins/outputs/timestream/timestream.go:471]: Yea, this can be a warning, updated.

File: plugins/outputs/timestream/timestream.go
Comment (lohithshetty) [reply plugins/outputs/timestream/timestream.go:460]: Updated

File: plugins/outputs/timestream/timestream.go
Comment (srebhan) [reply plugins/outputs/timestream/timestream.go:18]: It's still removed... :-)

**Comment by srebhan** [comment]
@lohithshetty I think you did not push the changes yet, did you? Let me know once you are ready for another review round...

File: plugins/outputs/timestream/timestream.go
Comment (lohithshetty) [reply plugins/outputs/timestream/timestream.go]: I just pushed the changes, thank you

**Comment by lohithshetty** [comment]
@[srebhan](https://github.com/srebhan) I pushed the changes, can you please take a look. Thank you.

**Comment by srebhan** [comment]
Hey @lohithshetty thanks for the push. This leaves us with the [question on why lines 123 - 126 are different in structure](https://github.com/influxdata/telegraf/pull/11385#discussion_r921983431)... Can you please comment on this?!

File: plugins/outputs/timestream/timestream.go
Comment (lohithshetty) [reply plugins/outputs/timestream/timestream.go:126]: when ingesting in MappingModeSingleTable and UseMultiMeasureRecords is enabled, we use measurementName ( from line protocol ) as multiMeasure name in timestream, where as in MappingModeMultiTable mode ( data is ingested to multiple tables ) and UseMultiMeasureRecords is enabled, we ask for measureName(MeasureNameForMultiMeasureRecords) that needs to be used, this is due to the fact that measurementName(line protocol) is used as timestream table name. Hope that clarifies, Readme has some details about different combinations for these configs.

Comment (lohithshetty) [reply plugins/outputs/timestream/timestream.go:126]: @nirmeshk Please add any details I might have missed here.

**Comment by lohithshetty** [comment]
Hi @[srebhan](https://github.com/srebhan), would appreciate your help in reviewing and closing this PR. Thank you.

File: plugins/outputs/timestream/timestream.go
Comment (srebhan) [reply plugins/outputs/timestream/timestream.go:160]: Fair enough. It would be good if you can add this explanation as a comment to the code...

**Comment by srebhan** [comment]
@lohithshetty beside my comment, can you please run `make docs` and push the changes to fix the unit tests!?

File: plugins/outputs/timestream/README.md
Comment (srebhan) [inline plugins/outputs/timestream/README.md:85]: All the changes in this section should happen in `sample.conf`. Afterwards, please run `make docs` to update the README.md file with your changes to the `sample.conf` file.

Comment (lohithshetty) [reply plugins/outputs/timestream/README.md:85]: Thanks @[srebhan](https://github.com/srebhan), I've updated sample.conf.

**Comment by lohithshetty** [comment]
Hi @[srebhan](https://github.com/srebhan), Thanks for reviewing the pull request, I've made the changes as per your suggestion. Currently circle-ci failing for mac for plugin plugins/outputs/graylog ( not related to this pull request ). Appreciate merging this changes. Thank you.

**Comment by kingston88** [comment]
@srebhan Are we good to merge this change this week?

File: plugins/outputs/timestream/sample.conf
Comment (reimda) [inline plugins/outputs/timestream/sample.conf:92]: Hi, thanks for updating the PR after reviews! I have another small one:

Could we move this section out of the toml and into README.md? The readme is the best place to put examples and longer sections like this. The toml comments should only be reminders of how to use specific variables.

We could leave a line like "see plugin readme for details" if you're worried people won't know to look in the readme.

Otherwise the PR looks good to me.

Comment (lohithshetty) [reply plugins/outputs/timestream/sample.conf:92]: Hi, 
Thanks for the suggestion, it makes sense. I just moved it under README.
readme_linter is failing for a different plugin, hope that doesn't block merging this PR.

**Review by reimda — APPROVED** [review]

**Comment by kingston88** [comment]
@reimda @srebhan Can we merge this change?

**Comment by reimda** [comment]
> @reimda @srebhan Can we merge this change?

We merge after two review approvals, so it's close. It looks like there is one outstanding item from srebhan's review that needs to be taken care of. Just adding some comments?

**Comment by lohithshetty** [comment]
> 

@reimda @srebhan  
README has a section which explains different modes with examples ( refer to `Mapping data from Influx to Timestream` section in README). Please let me know if you think more explanation/comments are required.

File: plugins/outputs/timestream/README.md
Comment (srebhan) [inline plugins/outputs/timestream/README.md:239]: You might want to use markdown tables for all those tables...

File: plugins/outputs/timestream/timestream.go
Comment (srebhan) [reply plugins/outputs/timestream/timestream.go:160]: The comment is still missing.

File: plugins/outputs/timestream/timestream_test.go
Comment (srebhan) [inline plugins/outputs/timestream/timestream_test.go:116]: Please split this
```suggestion
	err := invalidConfigMultiMeasureSingleTableMode.Connect()
	require.ErrorContains(r, err, "MeasureNameForMultiMeasureRecords")
```

**Review by srebhan — COMMENTED** [review]
Thanks @lohithshetty for the update! There are some small things here and there, but only formatting mostly. Can you please care for the [comment documenting why the structure differs](https://github.com/influxdata/telegraf/pull/11385#discussion_r931278842)?

File: plugins/outputs/timestream/README.md
Comment (lohithshetty) [reply plugins/outputs/timestream/README.md:239]: sure

File: plugins/outputs/timestream/timestream.go
Comment (lohithshetty) [reply plugins/outputs/timestream/timestream.go:257]: This adds no useful information to our customers and we had customers complaining high CPU utilization on their environment, after profiling we found this debug print causing issues as it takes in the entire writeRecordsInput to print.

File: plugins/outputs/timestream/timestream.go
Comment (lohithshetty) [reply plugins/outputs/timestream/timestream.go:495]: that's fair

File: plugins/outputs/timestream/timestream.go
Comment (lohithshetty) [reply plugins/outputs/timestream/timestream.go:289]: yep, makes sense.

File: plugins/outputs/timestream/timestream.go
Comment (srebhan) [reply plugins/outputs/timestream/timestream.go:257]: Ok. In this case removing the comment is a good thing.

**Review by srebhan — APPROVED** [review]
Looks good to me. Thanks for your contribution @lohithshetty!

**Comment by hadiel05** [comment]
Is this released? tested with this tag telegraf:1.24.3 and seems like its not.

**Comment by powersj** [comment]
Anything tagged a new "feature" is released in the next minor version. This should go out in v1.25.0

**Pull request merged**
